### ISSUE TITLE
PCNT units 4-7 not available on esp32 (IDFGH-4547)

### ISSUE THREAD

**Issue by TinkerBear** [espressif/esp-idf#6366]
## Environment

- Development Kit:      ESP32-DevKitC
- Kit version (for WroverKit/PicoKit/DevKitC): v4
- Module or chip used:  ESP32-WROOM-32D
- IDF version (run ``git describe --tags`` to find it): v4.2
- Build System:         idf.py
- Compiler version (run ``xtensa-esp32-elf-gcc --version`` to find it): 8.4.0
- Operating System:     macOS
- (Windows only) environment type: [MSYS2 mingw32|ESP Command Prompt|Plain Command Prompt|PowerShell].
- Using an IDE?: VSCode
- Power Supply:         USB

## Problem Description

PCNT units 4-7 are not available on esp32 after changes for esp32s2.  Number of units is defined in esp-idf/components/soc/soc/esp32/include/soc/pcnt_caps.h as SOC_PCNT_UNIT_NUM, which affects enum constants in esp-idf/components/soc/include/hal/pcnt_types.h

First problem: No other header file includes pcnt_caps.h, and it is not mentioned in documentation, so client-side code will never have the constants for all 8 units available.

Second problem: esp-idf/components/driver/pcnt.c includes pcnt_caps.h, but only after pcnt_types.h, so it has no effect, and the compiled code will not support units 4-7.  ("soc/pcnt_caps.h" must move to before "driver/pcnt.h".)

### Expected Behavior

That PCNT units 4-7 work as usual on esp32

### Actual Behavior

Compile or runtime errors for units 4-7

### Steps to reproduce

1) Open examples/peripherals/pcnt/main/pcnt_example_main.c
2) Replace "PCNT_UNIT_0" with "PCNT_UNIT_4"
3) Attempt to compile
4) Compile fails

// If possible, attach a picture of your setup/wiring here.

### Code to reproduce this issue

```cpp
PCNT_UNIT_4
```
// If your code is longer than 30 lines, [GIST](https://gist.github.com) is preferred.

## Debug Logs

```
Debug log goes here, should contain the backtrace, as well as the reset source if it is a crash.
Please copy the plain text here for us to search the error log. Or attach the complete logs but leave the main part here if the log is *too* long.
```

## Other items if possible

- [ ] sdkconfig file (attach the sdkconfig file from your project folder)
- [ ] elf file in the ``build`` folder (**note this may contain all the code details and symbols of your project.**)
- [ ] coredump (This provides stacks of tasks.)

**Comment by Alvin1Zhang** [comment]
Thanks for reporting.

**Comment by ESP-Marius** [comment]
@TinkerBear This has been merged into release/v4.2 with 10f16c5d727c548a2e07 and should be available the next time we sync with Github. The issue will be automatically updated when this happens.

I've also merged a similar commit to master which explicitly includes `x_caps.h` a bunch of places. 

Thanks for reporting this!

**Comment by Alvin1Zhang** [comment]
Thanks for reporting, the fix is available at https://github.com/espressif/esp-idf/commit/5e9783f4a6ffd96e016a6654a7bce2d12c5f6751, feel free to reopen. Thanks.
